src/providers.rs

```diff
@@ -16,232 +15,204 @@ use crate::{
-pub const BLOCKPI_OPTIMISM_MAINNET_HOSTNAME: &str = "optimism.blockpi.network";
-pub const PUBLICNODE_OPTIMISM_MAINNET_HOSTNAME: &str = "optimism-rpc.publicnode.com";
-pub const LLAMA_ETH_MAINNET_HOSTNAME: &str = "eth.llamarpc.com";
-pub const LLAMA_ARBITRUM_ONE_HOSTNAME: &str = "arbitrum.llamarpc.com";
-pub const LLAMA_BASE_MAINNET_HOSTNAME: &str = "base.llamarpc.com";
-pub const LLAMA_OPTIMISM_MAINNET_HOSTNAME: &str = "optimism.llamarpc.com";
+pub const PROVIDERS: &[Provider] = &[
+    Provider {
+        provider_id: 0,
+        chain_id: ETH_MAINNET_CHAIN_ID,
+        url_pattern: "https://cloudflare-eth.com/v1/mainnet/{API_KEY}",
+        header_patterns: &[],
+        service: Some(RpcService::EthMainnet(EthMainnetService::Cloudflare)),
+    },
+    Provider {
+        provider_id: 1,
+        chain_id: ETH_MAINNET_CHAIN_ID,
+        url_pattern: "https://rpc.ankr.com/eth/{API_KEY}",
+        header_patterns: &[],
+        service: Some(RpcService::EthMainnet(EthMainnetService::Ankr)),
+    },
+    Provider {
+        provider_id: 2,
+        chain_id: ETH_MAINNET_CHAIN_ID,
+        url_pattern: "https://ethereum-rpc.publicnode.com",
+        header_patterns: &[],
+        service: Some(RpcService::EthMainnet(EthMainnetService::PublicNode)),
+    },
+    Provider {
+        provider_id: 3,
+        chain_id: ETH_MAINNET_CHAIN_ID,
+        url_pattern: "https://ethereum.blockpi.network/v1/rpc/{API_KEY}",
+        header_patterns: &[],
+        service: Some(RpcService::EthMainnet(EthMainnetService::BlockPi)),
+    },
+    Provider {
+        provider_id: 4,
+        chain_id: ETH_SEPOLIA_CHAIN_ID,
+        url_pattern: "https://rpc.sepolia.org",
+        header_patterns: &[],
+        service: None,
+    },
+    Provider {
+        provider_id: 5,
+        chain_id: ETH_SEPOLIA_CHAIN_ID,
+        url_pattern: "https://rpc.ankr.com/eth_sepolia/{API_KEY}",
+        header_patterns: &[],
+        service: Some(RpcService::EthSepolia(EthSepoliaService::Ankr)),
+    },
+    Provider {
+        provider_id: 6,
+        chain_id: ETH_SEPOLIA_CHAIN_ID,
+        url_pattern: "https://ethereum-sepolia.blockpi.network/v1/rpc/{API_KEY}",
+        header_patterns: &[],
+        service: Some(RpcService::EthSepolia(EthSepoliaService::BlockPi)),
+    },
+    Provider {
+        provider_id: 7,
+        chain_id: ETH_SEPOLIA_CHAIN_ID,
+        url_pattern: "https://ethereum-sepolia-rpc.publicnode.com",
+        header_patterns: &[],
+        service: Some(RpcService::EthSepolia(EthSepoliaService::PublicNode)),
+    },
+    Provider {
+        provider_id: 8,
+        chain_id: ETH_MAINNET_CHAIN_ID,
+        url_pattern: "https://eth-mainnet.g.alchemy.com/v2/{API_KEY}",
+        header_patterns: &[],
+        service: Some(RpcService::EthMainnet(EthMainnetService::Alchemy)),
+    },
+    Provider {
+        provider_id: 9,
+        chain_id: ETH_SEPOLIA_CHAIN_ID,
+        url_pattern: "https://eth-sepolia.g.alchemy.com/v2/{API_KEY}",
```

Comment:
nit: for all the Alchemy providers I would use [authorization headers](https://docs.alchemy.com/docs/how-to-use-api-keys-in-http-headers), which is actually best-practice and reduce potential of leaking the key (typically servers have a "tendency" to log the full URL, while not so much for Authorization headers). Unfortunately BlockPi doesn't support this, so we have to do both. Also ok if you want to that in a separate PR (then maybe add a TODO).

Reply:
Sounds good. Updated the Alchemy providers to use authorization headers.